libs/vm-modules/src/ml/model/model.cpp

```diff
@@ -89,233 +123,65 @@ Ptr<VMModel> VMModel::Constructor(VM *vm, TypeId type_id,
+  auto const optimiser_type = DeduceOptimizer(optimiser->string());
   model_->Compile(optimiser_type, loss_type);
 }
 
 void VMModel::CompileSimple(fetch::vm::Ptr<fetch::vm::String> const &        optimiser,
                             fetch::vm::Ptr<vm::Array<math::SizeType>> const &in_layers)
 {
-  // construct the model with the specified layers
-  auto                        n_elements = in_layers->elements.size();
-  std::vector<math::SizeType> layers(n_elements);
-  for (std::size_t i = 0; i < n_elements; ++i)
+  OptimiserType optimiser_type = DeduceOptimizer(optimiser->string());
+  if (optimiser_type != OptimiserType::ADAM)
   {
-    layers.at(i) = in_layers->elements.at(i);
+    throw std::runtime_error(
+        R"(Wrong optimiser, a "Simple" model can use only "adam", while given : )" +
+        optimiser->string());
   }
 
+  auto BuildLayersFromInput = [&in_layers]() {
+    auto const n_elements = in_layers->elements.size();
+
+    std::vector<math::SizeType> layers(n_elements);
+    for (std::size_t i = 0; i < n_elements; ++i)
+    {
+      layers.at(i) = in_layers->elements.at(i);
+    }
+    return layers;
+  };
+
   switch (model_category_)
   {
   case (ModelCategory::REGRESSOR):
-  {
-    model_ = std::make_shared<fetch::ml::model::DNNRegressor<TensorType>>(*model_config_, layers);
+    model_ = std::make_shared<fetch::ml::model::DNNRegressor<TensorType>>(*model_config_,
+                                                                          BuildLayersFromInput());
     break;
-  }
+
   case (ModelCategory::CLASSIFIER):
-  {
-    model_ = std::make_shared<fetch::ml::model::DNNClassifier<TensorType>>(*model_config_, layers);
+    model_ = std::make_shared<fetch::ml::model::DNNClassifier<TensorType>>(*model_config_,
+                                                                           BuildLayersFromInput());
     break;
-  }
-  default:
-  {
-    throw std::runtime_error("speicified model type does not take layers on compilation");
-  }
-  }
 
-  // set up the optimiser and compile
-  fetch::ml::OptimiserType optimiser_type;
-  if (optimiser->string() == "adam")
-  {
-    optimiser_type = fetch::ml::OptimiserType::ADAM;
-  }
-  else
-  {
-    throw std::runtime_error("invalid optimiser");
+  default:
+    throw std::runtime_error(
+        "Only REGRESSOR and CLASSIFIER model types take layers on compilation!");
   }
 
   model_->Compile(optimiser_type);
 }
+
+// TODO(VH): Why the method is called "Fit" while it does only 1 epoch training?
 void VMModel::Fit(vm::Ptr<VMTensor> const &data, vm::Ptr<VMTensor> const &labels,
                   fetch::math::SizeType const &batch_size)
 {
-  // prepare dataloader
-  auto dl = std::make_unique<TensorDataloader>();
-  dl->SetRandomMode(true);
-  dl->AddData({data->GetTensor()}, labels->GetTensor());
-  model_->SetDataloader(std::move(dl));
+  auto data_loader = std::make_unique<TensorDataloader>();
```

Comment:
here is an example where auto is preferred because the Type is spelled out on the right hand side. so this one is fine.